docs/technical-design/contract-rate-refactor-relationships.md

```diff
@@ -0,0 +1,163 @@
+A.0, 1.0, 2020-01-01, null, false       
+A.0, 2.0, 2020-01-01, null, false        // this block is the initial state we enter. 
+B.0, 1.0, 2020-01-01, null, false       
+C.0, 2.0, 2020-01-01, null, false 
+```      
+
+then, we insert and submit a new rate, associated with A and B
+
+```
+A.0, 3.0, 2020-01-02, null, false
+B.0, 3.0, 2020-01-02, null, false
+```
+
+now if we unlock, update A, and resumit, creating a new revision, we add a new set of A relationships
+
+```
+A.1, 1.0, 2020-01-03, null, false
+A.1, 2.0, 2020-01-03, null, false
+A.1, 3.0, 2020-01-03, null, false
+```
+...and invalidate the old ones (these update previous rows) leading us to this table:
+
+```
+A.0, 1.0, 2020-01-01, 2020-01-03, false
+A.0, 2.0, 2020-01-01, 2020-01-03, false
+B.0, 1.0, 2020-01-01, null, false       
+C.0, 2.0, 2020-01-01, null, false 
+A.0, 3.0, 2020-01-02, 2020-01-03, false
+B.0, 3.0, 2020-01-02, null, false
+A.1, 1.0, 2020-01-03, null, false
+A.1, 2.0, 2020-01-03, null, false
+A.1, 3.0, 2020-01-03, null, false
+```
+
+### Version history
+Now, it's not super straightforward, but this is enough information to construct a sensible history. 
+
+For instance, we can try and determine the history for rate 1.0, by selecting all rows that have 1.0 as the rate revision, grouping by the valid_after timestamp. 
+
+```
+A.0, 1.0, 2020-01-01, 2020-01-03
+B.0, 1.0, 2020-01-01, null       
+A.1, 1.0, 2020-01-03, null
+```
+This is a toy example but you can see how you could construct a history like this:
+
+```
+2020-01-01: 
+    1.0 -> A.0
+    1.0 -> B.0
+2020-01-03:
+    1.0 -> A.1
+    1.0 -> B.0
+```
+
+This join table strategy allows us to maintain a fully versioned history of revisions to contracts, rate, and the associations between them. 
+
+Continuing the above example, let’s say we resubmit contract B and remove the link to rate 1. 
+
+We’ll have new entries for the rates, including the one removed:
+
+```
+B.1, 1.0, 2020-01-04, 2020-01-04, true
+B.1, 3.0, 2020-01-04, null, false
+```
+And we’ll update the old relations to be invalid
+
+```
+B.0, 1.0, 2020-01-01, 2020-01-04, false
+B.0, 3.0, 2020-01-02, 2020-01-04, false
+```
+
+This will give us a new entry in our rate 1 history, and we can attribute it to the correct submission because we have the isRemoved bit:
+```
+2020-01-01: – initial submit of rate 1
+    1.0 -> A.0
+    1.0 -> B.0
+2020-01-03: – contract A resubmitted
+    1.0 -> A.1
+    1.0 -> B.0
+2020-01-04: – contract B resubmitted
+    1.0 -> A.1
+```
+
+## Migration plan summary
+
+Existing `HealthPlanPackage` submissions will be converted into Contract and Rates in the db. Since we will be working on the migration work behind a feature flag, we can write a migrator that runs repeatedly, taking the old protobuf encoded FormData and moving that data into our new `Contract[Revision]` and `Rate[Revision]` tables. This way we can test and refine our migration before we actually switch over to using the new data. 
```

Comment:
I would like to see this section be even more explicit saying _yes  we are fully deleting health plan package tables after this change._ 

I feel like that detail could be missed right now. 

something like - this is a full schema migration where we will first build a new parallel set of tables but after we migrate the submission data, we end the use of the current `HealthPlanPackage` tables in favor of `Contract` and `Rates` and their related tables going forward.